add support for isort and --fix to spack style

cosmicexplorer opened the pull request:
### Problem

#23947 adds `flake8` import style checking, which is awesome. It would be great however if we could force `spack style` to fix import ordering itself. `isort` allows this and is compatible with `black`.

### Solution
- Add `run_isort()` to `style.py` to execute isort in `spack style`, and add a [`pyproject.toml`](https://pycqa.github.io/isort/docs/configuration/config_files/) configuration file.
    - See this page on [custom sections](https://pycqa.github.io/isort/docs/configuration/custom_sections_and_ordering/) and this page on [isort options](https://pycqa.github.io/isort/docs/configuration/options/) to understand the content of this config file.
- Add `isort` to our dependencies in `unit_tests.yaml`.
- Add a `--fix` argument to `spack style`, which configures isort and black to edit files on the local filesystem instead of printing out a diff.
- Add the `--show-error-codes` argument to mypy so that the user can easily see how to disable the warning with a focused `# type: ignore[<error-code>]` comment.
    - See [mypy's documentation on error codes](https://mypy.readthedocs.io/en/stable/error_codes.html).

### Result
We can automatically fix import sorting errors identified via #23947 with isort, and we can automatically apply the black formatter as well with the `--fix` option to `spack style`!

cosmicexplorer committed:
style: add support for `isort` and `--fix`

cosmicexplorer commented on lib/spack/spack/cmd/style.py:
```diff
@@ -33,7 +33,8 @@ def grouper(iterable, n, fillvalue=None):
     args = [iter(iterable)] * n
-    return zip_longest(*args, fillvalue=fillvalue)
+    for group in zip_longest(*args, fillvalue=fillvalue):
+        yield filter(None, group)
```
`filter(None, x)` is the same as `filter(lambda e: e is not None, x)`, which we do in every method that calls this iterator, so it seemed reasonable to move it up here.

cosmicexplorer commented on lib/spack/spack/cmd/style.py:
```diff
@@ -239,7 +252,7 @@ def run_mypy(file_list, args):
 
     print_tool_header("mypy")
-    mpy_args = ["--package", "spack", "--package", "llnl"]
+    mpy_args = ["--package", "spack", "--package", "llnl", "--show-error-codes"]
```
Not related to this change, but makes it significantly easier to know how to disable a mypy error.

trws replied:
Agreed, should have done this from the start.

adamjstewart replied:
We should move these to `.mypy.ini`. In fact, we should combine `.mypy.ini` into `pyproject.toml` so that we have one fewer top-level config file.

adamjstewart replied:
It looks like `pyproject.toml` support was added in `mypy>=0.900` so we should update the version requirement too.

tgamblin committed:
style: Move isort configuration to pyproject.toml

- [x] Remove flake8-import-order checks, as we only need isort for this
- [x] Clean up configuration and requirements

tgamblin committed:
style: clean up and restructure `spack style` command

This consolidates code across tools in `spack style` so that each
`run_<tool>` function can be called indirecty through a dictionary
of handlers, and os that checks like finding the executable for the
tool can be shared across commands.

- [x] rework `spack style` to use decorators to register tools
- [x] define tool order in one place in `spack style`
- [x] fix python 2/3 issues to Get `isort` checks working
- [x] make isort error regex more robust across versions
- [x] remove unused output option
- [x] change vestigial `TRAVIS_BRANCH` to `GITHUB_BASE_REF`
- [x] update completion

tgamblin commented:
@adamjstewart I made this work on `develop` and will submit a separate PR to fix everything.  See what you think of the changes.

I also reworked the output for `spack style`

adamjstewart commented on pyproject.toml:
```diff
@@ -0,0 +1,17 @@
+known_llnl = "llnl"
+src_paths = "lib"
+force_sort_within_sections = false
+line_length = 79
```
Shouldn't this be 88 (the default for `profile = 'black'`)?

tgamblin replied:
I guess if we're switching to black anyway we should go ahead and do it. Removed.

adamjstewart commented on pyproject.toml:
```diff
@@ -0,0 +1,17 @@
+force_sort_within_sections = false
+line_length = 79
+honor_noqa = true
+use_parentheses = true
```
This can be removed, it's implied by `profile = 'black'`

adamjstewart replied:
isort also has `color_output` and `skip_gitignore` options we could use.

tgamblin replied:
done

adamjstewart commented on pyproject.toml:
```diff
@@ -0,0 +1,17 @@
+known_archspec = "archspec"
+known_llnl = "llnl"
+src_paths = "lib"
+force_sort_within_sections = false
```
What does this imply? Don't we want to sort alphabetically within sections?

adamjstewart commented on lib/spack/spack/cmd/style.py:
```diff
@@ -306,15 +333,21 @@ def prefix_relative(path):
             file_list = changed_files(args.base, args.untracked, args.all)
         print_style_header(file_list, args)
+        if args.isort:
+            # We run isort before flake8, assuming that #23947 with flake8-import-order
+            # is also in, to allow flake8 to double-check the result of executing isort,
+            # if --fix was provided.
```
We should run both `black` and `isort` before `flake8` since they may fix things that `flake8` will complain about.

tgamblin replied:
consolidated

adamjstewart commented on lib/spack/spack/cmd/style.py:
```diff
@@ -252,20 +263,41 @@ def run_mypy(file_list, args):
+    print_tool_header("isort")
+    check_fix_args = () if args.fix else ("--check", "--diff")
+
+    pat = re.compile("ERROR: (.*) Imports are incorrectly sorted and/or formatted")
+    replacement = "ERROR: {0} Imports are incorrectly sorted and/or formatted"
```
How stable is this error message across multiple versions of isort?

tgamblin replied:
changed it to a shorter version but that string has been there for a while -- even in the test suite.

adamjstewart reviewed (commented):
I agree with dropping `flake8-import-order` in favor of `isort` as `isort` seems more powerful and more popular. It is possible to use both (you can write your own custom style for `flake8-import-order`) but I don't see any advantage in that. It's already hard enough to ensure that `flake8` and `black` agree on style; no need to add both `flake8-import-order` and `isort` to the mix. The only annoyance is that `flake8` will bring in any plugins it can find, including `flake8-import-order` which is in my path. We could disable this explicitly, but I might just stop using it for my projects.

adamjstewart commented on .github/workflows/unit_tests.yaml:
```diff
@@ -39,7 +39,7 @@ jobs:
     - name: Install Python packages
       run: |
-        pip install --upgrade pip six setuptools flake8 flake8-import-order mypy>=0.800 black types-six types-python-dateutil
+        pip install --upgrade pip six setuptools flake8 isort mypy>=0.800 black types-six types-python-dateutil
```
```suggestion
        pip install --upgrade pip six setuptools flake8 isort>=4.3.5 mypy>=0.800 black types-six types-python-dateutil
```
This is the first version of isort that supports pyproject.toml. Not sure if there are any newer features we're using, but an incorrect minimum version is better than no minimum version.

tgamblin replied:
done

adamjstewart commented on lib/spack/spack/cmd/style.py:
```diff
@@ -252,20 +263,41 @@ def run_mypy(file_list, args):
         return 1
 
     print_tool_header("black")
+    check_fix_args = () if args.fix else ("--check", "--diff", "--color")
```
`--color` could be put in the `pyproject.toml`

tgamblin replied:
it doesn't matter because it'll still decide to decolorize it when we read the output.

adamjstewart replied:
Color output works for me in GitHub Actions.

adamjstewart commented on pyproject.toml:
```diff
@@ -0,0 +1,17 @@
+known_archspec = "archspec"
+known_llnl = "llnl"
+src_paths = "lib"
+force_sort_within_sections = false
```
This is the default, so we could remove it.

tgamblin replied:
done

adamjstewart commented on .github/workflows/unit_tests.yaml:
```diff
@@ -39,7 +39,7 @@ jobs:
     - name: Install Python packages
       run: |
-        pip install --upgrade pip six setuptools flake8 flake8-import-order mypy>=0.800 black types-six types-python-dateutil
+        pip install --upgrade pip six setuptools flake8 isort>=4.3.5 mypy>=0.800 black types-six types-python-dateutil
```
```suggestion
        pip install --upgrade pip six setuptools flake8 isort[colors]>=4.3.5 mypy>=0.800 black types-six types-python-dateutil
```
This makes sure that colorama is installed too.

tgamblin replied:
we don't require colorama though...

adamjstewart replied:
If you add `color_output = true` to `pyproject.toml` and add `[colors]` then it will color the diffs.

tgamblin replied:
It has no option to colorize output that goes to a non-tty (i.e. the pipe), so we can't pass it through without more work.  The flags here will have no effect, at least for isort.

adamjstewart commented on lib/spack/spack/cmd/style.py:
```diff
@@ -134,6 +138,19 @@ def setup_parser(subparser):
+        "--fix",
+        action="store_true",
+        default=False,
+        help="If the tool supports automatically editing file contents, do that.",
```
These help messages shouldn't be capitalized or have periods to match other help messages (including builtins from argparse)

adamjstewart reviewed (dismissed):
There are still a few things I would change (add color to isort/black), move mypy.ini to pyproject.toml, but those can happen in a follow-up PR.

tgamblin committed:
style: get close to full coverage of `spack style`

Previous tests of `spack style` didn't really run the tools --
they just ensure that the commands worked enough to get coverage.

This adds several real tests and ensures that we hit the corner
cases in `spack style`.  This also tests sucess as well as failure
cases.

tgamblin commented:
Sigh ... not sure how to replicate this macOS failure.

I'll keep at it.

tgamblin committed:
cvs tests: don't use dateutil at all

`dateutil.parser` was an optional dependency for CVS tests. It was failing on macOS
beacuse the dateutil types were not being installed, and mypy was failing *even when the
CVS tests were skipped*. This seems like it was an oversight on macOS --
`types-dateutil-parser` was not installed there, though it was on Linux unit tests.

It takes 6 lines of YAML and some weird test-skipping logic to get `python-dateutil` and
`types-python-dateutil` installed in all the tests where we need them, but it only takes
4 lines of code to write the date parser we need for CVS, so I just did that instead.

Note that CVS date format can vary from system to system, but it seems like it's always
pretty similar for the parts we care about.

- [x] Replace dateutil.parser with a simpler date regex
- [x] Lose the dependency on `dateutil.parser`

tgamblin commented:
Ok found the issue -- `mypy` was failing due to the *type stubs* for `dateutil.parser` not being installed in the macOS tests. Sigh.

`dateutil.parser` was added as an import for tests in #23212 to parse dates out of CVS log output for the mock CVS repo that the tests build.  It had to be made optional in #24484.  And here, it seems that even though it was a conditional import, `mypy` is a static tool and needs all the info everywhere.

It looks like there was some discussion of this in https://github.com/spack/spack/pull/23212#issuecomment-862533698, and the issue is that CVS dates can obnoxiously have different formats on different systems. I googled around and it looks like they can look like any of these:

```
date: 2021-07-07 02:43:33 -0700;  ...
date: 2021-07-07 02:43:33;  ...
date: 2021/07/07 02:43:33 -0700;  ...
date: 2021/07/07 02:43:33;  ...
```

So, slashes or dashes, with or without timezone. There never seems to be a sub-second component.

It takes 6 lines of YAML and some weird test-skipping logic to get `python-dateutil` and `types-python-dateutil` installed in all the tests where we need them, but it only takes 4 lines of code to write the minimal date parser we need for CVS, so I just did that instead.  This has the added benefit that the CVS tests run anywhere CVS is installed; not just where CVS and `dateutil` are installed.  And we no longer need to worry about `dateutil` compatibility across python versions.

FYI @alalazo, @eschnett, @scheibelp.

alalazo commented on lib/spack/spack/cmd/style.py:
```diff
@@ -32,39 +34,49 @@ def grouper(iterable, n, fillvalue=None):
     git_args = [
         # Add changed files committed since branching off of develop
-        ["diff", "--name-only", "--diff-filter=ACMR", range],
+        ["diff", "--name-only", "--diff-filter=ACMR", "--", range],
```
Reading the docs it seems that `--` is used to distinguish commit ref from paths,
```suggestion
        ["diff", "--name-only", "--diff-filter=ACMR", range],
```
so I think it should be omitted here?

tgamblin replied:
Nice catch. I had to do this to get things to work for some reason but I cannot reproduce now.  Deleted.

tgamblin commented:
ok we might actually get this in this time -- thanks @alalazo for the late catch

adamjstewart commented on lib/spack/spack/cmd/style.py:
```diff
@@ -113,12 +135,6 @@ def setup_parser(subparser):
         action="store_true",
         help="check all files, not just changed files",
     )
-    subparser.add_argument(
-        "-o",
-        "--output",
-        action="store_true",
-        help="send filtered files to stdout as well as temp files",
-    )
```
Should this be deprecated before removing?

tgamblin replied:
tempted to say no as I do not think anyone uses it for anything except for (maybe?) us.  Also it was already not working or doing anything.
